src/main/scala/org/graphframes/Pregel.scala

```diff
@@ -0,0 +1,236 @@
+   * aggregated messages in each round.
+   *
+   * @param colName the column name of initialized column in vertex Dataframe
+   * @param initialExpr The column expression used to initialize the column.
+   * @param updateAfterAggMsgsExpr The column expression used to update the column.
+   *                               Note that this sql expression can reference all
+   *                               vertex columns and an extra message column
+   *                               `Pregel.msgCol`.
+   */
+  def withVertexColumn(colName: String, initialExpr: Column,
+                       updateAfterAggMsgsExpr: Column): this.type = {
+    withVertexColumnList += Tuple3(colName, initialExpr, updateAfterAggMsgsExpr)
+    this
+  }
+
+  /**
+   * Set the message column. In each round of pregel, each triplet
+   * (src-edge-dst) will generate zero or one message and the message will
+   * be sent to the src vertex of this triplet.
+   *
+   * @param msgExpr The message expression. It is a sql expression and it
+   *                can reference all propertis in the triplet, in the way
+   *                `Pregel.src("src_col_name)`, `Pregel.edge("edge_col_name)`,
+   *                `Pregel.dst("dst_col_name)`
+   */
+  def sendMsgToSrc(msgExpr: Column): this.type = {
+    sendMsgs += Tuple2(Pregel.src("id"), msgExpr)
+    this
+  }
+
+
+  /**
+   * Set the message column. In each round of pregel, each triplet
+   * (src-edge-dst) will generate zero or one message and the message will
+   * be sent to the dst vertex of this triplet.
+   *
+   * @param msgExpr The message expression. It is a sql expression and it
+   *                can reference all propertis in the triplet, in the way
+   *                `Pregel.src("src_col_name)`, `Pregel.edge("edge_col_name)`,
+   *                `Pregel.dst("dst_col_name)`
+   */
+  def sendMsgToDst(msgExpr: Column): this.type = {
+    sendMsgs += Tuple2(Pregel.dst("id"), msgExpr)
+    this
+  }
+
+  /**
+   * Set the aggregation expression which used to aggregate messages which received
+   * by each vertex.
+   *
+   * @param aggExpr The aggregation expression, such as `sum(Pregel.msgCol)`
+   */
+  def aggMsgs(aggExpr: Column): this.type = {
+    aggMsgsCol = aggExpr
+    this
+  }
+
+  /**
+   * After set a series of things via above methods, then call this method to run
+   * pregel, and it will return the result vertex dataframe, which will include all
+   * updated columns in the final rounds of pregel.
+   *
+   * @return the result vertex dataframe
+   */
+  def run(): DataFrame = {
+    val sendMsgsColList = sendMsgs.toList.map { case (id, msg) =>
+      struct(id.as("id"), msg.as("msg"))
+    }
+
+    val initVertexCols = withVertexColumnList.toList.map { case (colName, initExpr, _) =>
+      initExpr.as(colName)
+    }
+    val updateVertexCols = withVertexColumnList.toList.map { case (colName, _, updateExpr) =>
+      when(Pregel.msg.isNotNull, updateExpr).otherwise(col(colName)).as(colName)
```

Comment:
Just tell users if a vertex doesn't receive messages, its msg col will be null. We shouldn't add logic that keep the col(colName) unchanged in such case.

Reply:
The logic of not updating the "rank" column when there is no msg received is not consistent with Pregel algorithm definition. For example, in page rank, a node without any incoming edge will lose the rank after one iteration (or the reset probability). We cannot keep its rank unchanged in that case. Leave it to users.